### PR TITLE
Broadcast Networking

### PR DESCRIPTION
**Pull request by adamelliot**
This PR introduction a broadcast network layer that allows for lamps to interact with the networking layer on a observation basis. IE you tell the network layer what you'd like to listen to.

Keep in mind most of this description is in the README in parts.

To allow for a decoupled and testable network layer I've done a bunch of decoupling of the app.

### First off the file structure

`src/` -  is where all the code for a lamp lives, this gets flashed to devices (in entirety, so any lamp could really switch to another)
`test/` -  is test cases, pretty much just the network module. Testing the app entry point and BLE layer is possible, but this was already getting large.
`src/*.py` - Just the bare bones boot files for the app, rest should live in the app folder. Makes things more portable
`src/app/network` - The networking module.
`src/app/lamp_core` - Core code that can be leveraged for all lamps. I think can do more house keeping here.
`src/app/utils` - Utility files not useful for all lamps, but handy to have around
`src/app/lamps` - Each lamp.  Each lamp will have it's own entry which is loaded from `app.py` the files live here.

### App Launch / Invoke Script / Loading Data on the ESP32s

The lamp loading mechanism works by utilizing the lamps filename as the lamp to load.  All `invoke` commands (or `inv` for short) that interact with the board should be followed by the file of the lamp in question.  eg: `inv run twinkle`.

### `lamp-mapping.json`

The `lamp-mapping.json` file allows to set a default pairing of lamps to devices so the `--port` argument can be dropped.

#### Flash and run the lamp on the target device
```
inv run gramp --port /dev/tty.usbserial-0246D45F
```

#### Flash the lamp to device
Note that flashing is a progressive operation. The build system will try to determine what files that need to be updated and only upload changed files since the last flash. If there is issues run `inv wipe --port [device]` first.
```
inv flash gramp --port /dev/tty.usbserial-0246D45F
```
#### Clean off all the code from 
```
inv wipe --port /dev/tty.usbserial-0246D45F
```
#### Run Test Suite
```
inv test
```

## Architecture of the Networking Layer

The primary complication or challenge of getting broadcast networking to work with BLE is the amount of data that can be sent in each payload is restricted to 32 bytes. This includes all the BLE headers and device name, which usually eats up 10 bytes.  To get around this restriction the network layer cycles the messages that should be shared with the other lamps.

General flow of networking.
1. Application starts up and launches the network monitoring thread.
2. Application launches specific lamp and hands initialized network object to the lamp.
3. Lamp can now register attributes it would like to tell the network about itself or send broadcast messages.

A lamp attribute is a property about the lamp, such as it's `BASE_COLOR`, `SHADE_COLOR` or version. These properties will not be repeated through the network but are always available about a specific lamp.

A broadcast message is some the lamp would like other lamps to participate in. Such as `BASE_OVERRIDE`.  These messages will be repeated as soon as any lamp sees the message. That way the should permeate the network. We should create some expectation that lamps respond to some of these messages to all for more rich interactions, but at the base line just adjusting to the broadcast colors would be fine.

`BroadcastMessage`s unlike `LampAttribute`s have a `ttl` so they will only last for a brief amount of time. The `ttl` also decays each time it hops to ensure we don't end up in lamp repeater loops.

## How to review this PR

The meat of the PR is the `network` folder. Start there, Using the test cases that parallel each file will help the reviewer follow what the expected behaviour is.

### CODE REVIEW

File: src/app/lamps/thing_broadcast.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from machine import Pin
+from ..lamp_core.base_lamp import BaseLamp
+from ..network.coding import Codes
+from ..colors import *
+
+class ThingBroadcast(BaseLamp):
```
Comment (jerrett) [inline src/app/lamps/thing_broadcast.py:6]: What makes this Thing specific (vs just "Broadcast" ?

File: src/app/lamp_core/base_lamp.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from ..network.coding import Codes
+from ..network.network import LampNetworkObserver
+
+class BaseLamp(LampNetworkObserver):
+    '''
+    Trival Lamp framework that real lamps will be based on.
+    Provides a simple interface for the rest of the system to build
+    up the lamp and does basic broadcasting about the lamp being created.
+    '''
+    def __init__(self, network, base_color, shade_color):
```
Comment (jerrett) [inline src/app/lamp_core/base_lamp.py:10]: where does `network` come from?

File: src/app/app.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        self._load_config()
+
+        self.network = LampNetwork(self.config['name'])
+        self._ble = LampBluetooth(self.network)
+
+        self.network.announce_attribute(Codes.VERSION, int(self.config['version']))
+
+        self._lamp_class = None
+        App.shared_app = self
```
[the remark this answers is not in the document]
Comment (adamelliot) [reply src/app/app.py:25]: Top level singleton so the `global` scoped lamps can interact with the `App`.

File: src/app/lamps/thing_broadcast.py (hunk 1 above)
Comment (adamelliot) [reply src/app/lamps/thing_broadcast.py:6]: This is for the `ESP32Thing`. It has a button on it which is non-standard. I mostly just use this to do broadcast tests.

File: src/app/lamp_core/base_lamp.py (hunk 2 above)
Comment (adamelliot) [reply src/app/lamp_core/base_lamp.py:10]: `network` is created in the `App` layer and injected to all the lamps.

File: src/app/lamps/thing_broadcast.py (hunk 1 above)
Comment (jerrett) [reply src/app/lamps/thing_broadcast.py:6]: aha

**Review by jerrett — APPROVED** [review]

**Pull request merged**